R/layout.R

```diff
@@ -483,16 +504,48 @@ col_width_grid_classes <- function(breaks, n_kids, n_cols = 12) {
 }
 
 
-col_width_best_fit <- function(kids, prefer_wider = FALSE) {
-  if (kids < 4) return(1)
-  if (kids <= 7) {
-    # sizes 4-7 are special cased to use (2 * kids) columns
-    return(if (prefer_wider) kids else 2)
+col_width_best_fit <- function(n, prefer_wider = FALSE, n_cols = NA) {
+  # If we're here, the user asked us to make the best choice possible about the
+  # number of columns, either by giving `col_widths = NA` or by using `NA` at
+  # a specific break point. The general idea is play with both the column
+  # widths and the `--bs-columns` variable (number of columns in the grid) to
+  # get decent results for a low number of items (1-7). At 7+ we use the
+  # 12-column grid and pick the factor that results in the fewest empty columns.
+
+  fit <- list(n_cols = n_cols)
+
+  # If n_cols is NA and n <= 7, best fit can adjust the n_cols for a better fit
+  if (is.na(fit$n_cols)) {
+    fit$n_cols <- if (n > 7) {
+      12
+    } else if (n > 3) {
+      n * 2
+    } else {
+      n
+    }
+
+    if (n < 4) {
+      fit$col_widths <- 1
+      return(fit)
+    }
+
+    if (n <= 7) {
+      # sizes 4-7 are special cased to use (2 * kids) columns
+      fit$col_widths <- if (prefer_wider) n else 2
+      return(fit)
+    }
+  } else if (n <= 3) {
+    # We're in fixed 12-column mode, manually pick for small numbers
+    fit$col_widths <- c(12, 6, 4)[n]
+    return(fit)
+  } else if (n == 5) {
+    fit$col_widths <- if (prefer_wider) 4 else 3
+    return(fit)
   }
```

Comment:
It'd be great to have some commentary about why these cases need to be handled differently than the "base case"

Reply:
Hmm, I was trying to cram a lot into "manually pick", yeah. They're handled differently because the best fit algorithm only picks from `c(2, 3, 4)` or `c(3, 4, 6)` and in the exception cases here we can make better choices. E.g. 1 item should always be a 12-unit column, 2 items should always be 6-unit columns, etc., but the best fit algorithm would pick something else. (Same for 5.)

What from the above would you find helpful in the comment?

Reply:
Writing out the above helped, I expanded the comment a bit in 1c872c4bdfbab577690a5b6d290f927352093e74

Reply:
Thanks, that does help